src/Init/System/Uri.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Lovett
+-/
+
+namespace System
+namespace Uri
+namespace UriEscape
+
+def hexDigitToNat (c : Char) : Option Nat :=
+  if '0' ≤ c ∧ c ≤ '9' then some (c.toNat - '0'.toNat)
+  else if 'a' ≤ c ∧ c ≤ 'f' then some (c.toNat - 'a'.toNat + 10)
+  else if 'A' ≤ c ∧ c ≤ 'F' then some (c.toNat - 'A'.toNat + 10)
+  else none
+
+def decodeUri (uri : String) : String := Id.run do
+  let mut decoded : ByteArray := ByteArray.empty
+  let len := uri.data.length
+  let mut pos : Nat := 0
+  while pos < len do
+    let c := uri.data[pos]!
+    if c == '%' && pos + 2 < len then
+      let h1 : Char := uri.data[pos + 1]!
+      if (h1 == '%') then
+        -- this is an escaped '%%' which should become a single percent.
+        decoded := decoded.push 37 -- %
+        pos := pos + 2
+      else
+         -- should have %HH where HH are hex digits, if they are not
+         -- valid hex digits then just repeat whatever sequence of chars
+         -- we found here.
+         let h2 := uri.data[pos + 2]!
+         let hc : Option UInt8 := match hexDigitToNat h1 with
+         | none => none
+         | some hd1 =>
+           match hexDigitToNat h2 with
+           | some hd2 => (hd1 * 16 + hd2).toUInt8
+           | none => none
+         decoded := match hc with
+         | some b => decoded.push b
+         | none => (decoded.push 37) ++ h1.toString.toUTF8 ++ h2.toString.toUTF8
+         pos := pos + 3
+    else
+      decoded := decoded.append c.toString.toUTF8
+      pos := pos + 1
+   return String.fromUTF8Unchecked decoded
```

Comment:
Instead of `ByteArray` and `fromUTF8Unchecked`, you could use `String.push` directly.

Reply:
String.push takes a Char (a decoded Unicode code point) not a byte, the % escaped values in a Uri are bytes, not Unicode code points.  For example:

```lean
def dump_string (s:String) : IO Unit := do
  for c in s.data do
    IO.println c.toNat

#eval dump_string "😵‍💫"
```
prints 3 Chars:
```
128565
8205
128171
```

Which are then converted to UTF-8 bytes and escaped in a Uri as `%3f%f0%9f%98%b5%e2%80%8d%f0%9f%92%ab%3d`

So to get this back into a string one has to create the 13 byte ByteArray from decoding these % escapes, then call fromUTF8Unchecked to decode that back into the 3 Chars Unicode codepoints.

Reply:
I see, I guess the alternative is redoing UTF-8 parsing in order to know when to call `push`. Not sure if we should do this in order to validate the input, or just trust it (the `Unchecked` call can definitely create invalid UTF-8 given a malformed URI-encoded string).

Reply:
Yes, you can encode illegal utf-8 sequences in the %HH sequence and get back random junk, for example:
```
#eval unescapeUri "test%FF%FF"  -- "testAA"
```